beamer/agent/chain.py

```diff
@@ -476,6 +482,139 @@ def maybe_invalidate(claim: Claim, context: Context) -> None:
+    timestamp: Timestamp, context: Context, target_chain_id: ChainId
+) -> bool:
+    # The internal memory should not be none because at this point
+    # the ChainUpdated event must have arrived, otherwise there would be no request
+    target_chain_finality = context.finality_periods[target_chain_id]
+    return int(time.time()) > timestamp + target_chain_finality
+
+
+def maybe_resolve_invalidation(claim: Claim, context: Context) -> bool:
+    request = context.requests.get(claim.request_id)
+
+    assert request is not None, "Active claim for non-existent request"
+
+    assert claim.invalidation_tx is not None, "Claim not invalidated"
+    assert claim.invalidation_timestamp is not None, "Claim not invalidated"
+
+    if not _timestamp_is_l1_finalized(
+        claim.invalidation_timestamp, context, request.target_chain_id
+    ):
+        return False
+
+    if not _l1_resolution_threshold_reached(claim, context):
+        return False
+
+    if claim.id in context.l1_invalidations:
+        return False
+
+    future = context.task_pool.submit(
+        run_relayer_for_tx,
+        context.config.base_chain_rpc_url,
+        context.target_rpc_url,
+        context.source_rpc_url,
+        context.config.account.key,
+        claim.invalidation_tx,
+    )
+
+    def on_future_done(f: Future) -> None:
+        try:
+            f.result()
+        except Exception as ex:
+            context.logger.error("L1 invalidation failed", ex=ex)
+        finally:
+            assert claim is not None, "Claim object missing"
+            del context.l1_invalidations[claim.id]
+
+    future.add_done_callback(on_future_done)
+    context.l1_invalidations[claim.id] = future
+
+    context.logger.info("Initiated L1 invalidation", request=request, claim=claim)
+
+    return True
+
+
+def maybe_resolve_request(claim: Claim, context: Context) -> bool:
+    request = context.requests.get(claim.request_id)
+
+    assert request is not None, "Active claim for non-existent request"
+
+    assert request.fill_tx is not None, "Request not yet filled"
+    assert request.fill_timestamp is not None, "Request not yet filled"
+
+    if not _timestamp_is_l1_finalized(request.fill_timestamp, context, request.target_chain_id):
+        return False
+
+    if not _l1_resolution_threshold_reached(claim, context):
+        return False
+
+    if request.id in context.l1_resolutions:
+        return False
+
+    future = context.task_pool.submit(
+        run_relayer_for_tx,
+        context.config.base_chain_rpc_url,
+        context.target_rpc_url,
+        context.source_rpc_url,
+        context.config.account.key,
+        request.fill_tx,
+    )
+
+    def on_future_done(f: Future) -> None:
+        try:
+            f.result()
+        except Exception as ex:
+            context.logger.error("L1 resolution failed", ex=ex)
+        finally:
+            assert request is not None, "Request object missing"
+            del context.l1_resolutions[request.id]
```

Comment:
hmm we have a race condition here. The relayer process will most likely be done before the event arrived at the agent. In this scenario we would run the relayer again, mistakenly.
Now it gets a bit tricky. If the relayer waits until he sees the event, we can be fine that the event is there. In this case we could add a flag (or just don't remove the value of the dict) 
Because then we can be sure that the agent will read the event eventually. 

Only if the process fails, we have to delete the value in order to allow another process to be started